### PR TITLE
[ENH] Implementing the `iTransformer` model in PTFv2.

### LINKED ISSUE #1899: [ENH] implementing `iTransformer` from `tslib`.
**Is your feature request related to a problem? Please describe.**
Implementing the `iTransformer` model in PTFv2. 

Basis: https://github.com/thuml/iTransformer/blob/main/model/iTransformer.py
Paper: https://arxiv.org/pdf/2310.06625

### PR DESCRIPTION
**Pull request by JATAYU000**
#### Reference Issues/PRs

Fixes #1899 

#### What does this implement/fix? Explain your changes.

Have started interfacing `iTransformer` in PTFv2, from the TSlib repository [thuml/iTransformer](https://github.com/thuml/iTransformer/blob/main/model/iTransformer.py)
Work in progress, would like to have suggestions on it.

#### What should a reviewer concentrate their feedback on?

- Current Implementation compliance with PTFv2

#### Did you add any tests for the change?
Not yet

#### PR checklist

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [ ] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

File: pytorch_forecasting/models/itransformer/submodules.py
Comment (PranavBhatP) [inline pytorch_forecasting/models/itransformer/submodules.py:1]: Conventionally for v2, all the layers of the models' architecture are present in the [layers](https://github.com/sktime/pytorch-forecasting/tree/main/pytorch_forecasting/layers) directory. Many of the layers you are using here can be directly imported from this directory, I see a lot of commonality. I would suggest only adding new layers  (if not already present in `layers`) as a subdirectory - `layers/_<layer-type>`. `sub_modules.py` is a v1 convention

File: pytorch_forecasting/models/itransformer/_itransformer_v2.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        # B N E -> B N E
+        # the dimensions of embedded time series has been inverted
+        enc_out, attns = self.encoder(enc_out, attn_mask=None)
+
+        # B N E -> B N S -> B S N
+        dec_out = self.projector(enc_out).permute(0, 2, 1)[
+            :, :, :N
+        ]  # filter covariates
+
+        if self.use_norm:
```
Comment (PranavBhatP) [inline pytorch_forecasting/models/itransformer/_itransformer_v2.py]: `self.use_norm` is not required in the model code since it will be handled by the D1/D2 layers. Normalization and denomalization need not be handled here. Simply return `dec_out`.

File: pytorch_forecasting/models/itransformer/_itransformer_pkg_v2.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        return {
+            "train": train_dataloader,
+            "val": val_dataloader,
+            "test": test_dataloader,
+            "data_module": train_datamodule,
+        }
+
+    @classmethod
+    def get_test_train_params(cls):
```
Comment (PranavBhatP) [inline pytorch_forecasting/models/itransformer/_itransformer_pkg_v2.py:36]: Can you add a few more test cases here?

**Review by PranavBhatP — CHANGES_REQUESTED** [review]
Thanks for the PR @JATAYU000 . I've dropped some comments on the PR.

File: pytorch_forecasting/models/itransformer/_itransformer_v2.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+from torch.optim import Optimizer
+
+from pytorch_forecasting.models.base._tslib_base_model_v2 import TslibBaseModel
+
+
+class iTransformer(TslibBaseModel):
+    """
+    An implementation of iTransformer model for v2 of pytorch-forecasting.
+
+    Parameters
```
Comment (PranavBhatP) [inline pytorch_forecasting/models/itransformer/_itransformer_v2.py:24]: Docstring for model hyperparameters is missing.

File: pytorch_forecasting/models/itransformer/submodules.py
Comment (JATAYU000) [reply pytorch_forecasting/models/itransformer/submodules.py:1]: I found some changes in some layers and let it be in `submodules` for now in draft pr, will fix this .

File: pytorch_forecasting/models/itransformer/_itransformer_v2.py (hunk 1 above)
Comment (JATAYU000) [reply pytorch_forecasting/models/itransformer/_itransformer_v2.py]: Oh Thank you for pointing that out.

**Comment by JATAYU000** [comment]
> I would suggest only adding new layers (if not already present in layers) as a subdirectory - layers/_<layer-type>

@PranavBhatP The `EncoderLayer` in `layers/_encoders/` requires `cross_attention`  , but `iTransformer` only needs `self_attention`.
Should I make `cross_attention` optional in the existing layer? or create a separate `EncoderLayer` for `iTransformer`?

**Comment by fkiraly** [comment]
re layers, I would do as follows:

* if exact same layer is available in `layers`, reuse it
* add new layers in `layers`
* if layer with modification is needed, add it as a separate layer
* optionally - this PR but also can be later PR - check if multiple similar layers can be "unified" in a single layer with more parameters

**Comment by PranavBhatP** [comment]
> optionally - this PR but also can be later PR - check if multiple similar layers can be "unified" in a single layer with more parameters

@fkiraly seems like a nice good first issue?

**Comment by PranavBhatP** [comment]
@JATAYU000 any hurdles with fixing issues in the PR?

**Comment by JATAYU000** [comment]
@PranavBhatP There are'nt any hurdles, I was busy with some other projects, I needed a review on the model and layer implementation since it updates the `TimeXer` layer to be a common TSLib layer, Have to add tests will do that.

File: pytorch_forecasting/layers/_encoders/_encoder.py [hunk 4]
```diff
@@ -2,39 +2,50 @@
...
-import torch
 import torch.nn as nn
-import torch.nn.functional as F
 
 
 class Encoder(nn.Module):
     """
-    Encoder module for the TimeXer model.
+    Encoder module for Tslib models.
     Args:
```
Comment (phoeenniixx) [inline pytorch_forecasting/layers/_encoders/_encoder.py:11]: please change this to numpydoc style docstring

File: pytorch_forecasting/layers/_encoders/_encoder_layer.py [hunk 5]
```diff
@@ -2,36 +2,34 @@
...
 import torch
 import torch.nn as nn
 import torch.nn.functional as F
 
 
 class EncoderLayer(nn.Module):
     """
-    Encoder layer for the TimeXer model.
+    Encoder layer for TsLib models.
     Args:
```
Comment (phoeenniixx) [inline pytorch_forecasting/layers/_encoders/_encoder_layer.py:13]: please change this to numpydoc style docstring

File: pytorch_forecasting/models/itransformer/_itransformer_pkg_v2.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        """Get test train params."""
+        from pytorch_forecasting.metrics import QuantileLoss
+
+        params = [
+            {},
+            dict(d_model=16, n_heads=2, e_layers=2, d_ff=64),
+            dict(
+                d_model=32,
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/itransformer/_itransformer_pkg_v2.py:44]: should we try one param with `output_attetion=True` as well? to cover all possibilities?

**Comment by agobbifbk** [comment]
This is misleading, timestamp are not processed by this layer :-) 
```python
enc_out = self.enc_embedding(x_enc, x_mark_enc)  # covariates (e.g timestamp) 
```
Moreover here I don't see any reference to the crossattention
```python
        self.encoder = Encoder(
            [
                EncoderLayer(
                    self_attention=AttentionLayer(
                        FullAttention(
                            False,
                            self.factor,
                            attention_dropout=self.dropout,
                            output_attention=True,
                        ),
                        self.d_model,
                        self.n_heads,
                    ),
                    d_model=self.d_model,
                    d_ff=self.d_ff,
                    dropout=self.dropout,
                    activation=self.activation,
                    output_attention=True,
                )
                for _ in range(self.e_layers)
            ],
            norm_layer=torch.nn.LayerNorm(self.d_model),
            output_attention=True,
        )
        if self.n_quantiles is not None:
            self.projector = nn.Linear(
                self.d_model, self.prediction_length * self.n_quantiles, bias=True
            )
        else:
            self.projector = nn.Linear(self.d_model, self.prediction_length, bias=True)
```

the Encoder layer has crossattention None by default
```
class EncoderLayer(nn.Module):
    def __init__(
        self,
        self_attention,
        cross_attention=None,
        d_model=512,
        d_ff=None,
        dropout=0.1,
        activation="relu",
        output_attention=False,
    ):
```
How the user can use this? Or I'm missing something?

**Comment by JATAYU000** [comment]
> This is misleading, timestamp are not processed by this layer :-)

The comment was cut, it was supposed to be `# covariates (e.g timestamp) are embedded as tokens`

> Moreover here I don't see any reference to the cross_attention

`iTransformer` does not require to pass `cross_attention` since it defaults to None.
when it is passed it is included in the block, defaulting it to None was to make the layer flexible to the layers which doesn't need `cross_atention` as in `iTransformer` or any other inverse models implemented from `TsLib` in the future

If there is any other reason to not to default to None? I did not understand the exact problem could you please  elaborate a bit more @agobbifbk
